### PR TITLE
Add ERC20 send support for faucet

### PR DESCRIPTION
**Pull request by webmaster128**

### CODE REVIEW

File: scripts/cosm/deploy_erc20.js [hunk 1]
```diff
@@ -118,7 +118,18 @@ async function main() {
       },
     ],
   };
-  for (const initMsg of [initMsgAsh, initMsgBash]) {
+  const initMsgCash = {
```
Comment (ethanfrey) [inline scripts/cosm/deploy_erc20.js]: Ah, here the famous 18 decimals... trying to overflow :wink:

File: packages/bcp/src/cosmwasmcodec.spec.ts [hunk 2]
```diff
@@ -1,64 +1,128 @@
...
+        chainId: chainId,
+        sender: "cosmos1txqfn5jmcts0x0q7krdxj8tgf98tj0965vqlmq" as Address,
+        recipient: "cosmos1dddd" as Address,
+        memo: "My first BASH payment",
+        amount: {
+          fractionalDigits: 0,
+          quantity: "345",
+          tokenTicker: "BASH" as TokenTicker,
+        },
+        fee: {
```
Comment (ethanfrey) [inline packages/bcp/src/cosmwasmcodec.spec.ts]: wow, almost 1/3 of this tx is just setting the fee and memo, which we would like to make automatic. Maybe use `defaultFee` for the tests. I definitely need to think about how to configure this better in production.

(I'm really looking at the API and what we can simplify in a basic API, without removing the power of the full API - I really think we need some wrapper class which makes lots of assumptions for intro devs, then they can pull off the wrapper and grab all the fields when they need it)

File: packages/bcp/src/cosmwasmconnection.spec.ts [hunk 3]
```diff
@@ -433,5 +458,47 @@ describe("CosmWasmConnection", () => {
 
       connection.disconnect();
     });
+
+    it("can send ERC20 tokens", async () => {
```
Comment (ethanfrey) [inline packages/bcp/src/cosmwasmconnection.spec.ts]: :smile:

File: packages/bcp/src/cosmwasmconnection.ts [hunk 4]
```diff
@@ -172,26 +171,31 @@ export class CosmWasmConnection implements BlockchainConnection {
...
           return {
             fractionalDigits: erc20.fractionalDigits,
-            quantity: response.balance,
+            quantity: normalizedBalance,
             tokenTicker: erc20.ticker as TokenTicker,
           };
         },
       ),
     );
+    const nonZeroErc20Amounts = erc20Amounts.filter(amount => amount.quantity !== "0");
```
Comment (ethanfrey) [inline packages/bcp/src/cosmwasmconnection.ts]: This is both cool, and expensive if we have more than say 5-10 erc20 defined. (Have you seen the list on myetherwallet... it is like 100+). Anyway, if the problem we have is too many tokens on the chain, then we are doing good :joy:

File: packages/bcp/src/encode.ts [hunk 5]
```diff
@@ -79,37 +77,83 @@ export function encodeFullSignature(fullSignature: FullSignature): types.StdSign
...
-    value: {
-      msg: [
-        {
-          type: "cosmos-sdk/MsgSend",
-          value: {
-            from_address: tx.sender,
-            to_address: tx.recipient,
-            amount: [encodeAmount(tx.amount, tokens)],
+
+  const matchingBankToken = bankTokens.find(t => t.ticker === tx.amount.tokenTicker);
```
Comment (ethanfrey) [inline packages/bcp/src/encode.ts]: Very nice!

File: packages/faucet/src/faucet.spec.ts [hunk 6]
```diff
@@ -81,7 +75,34 @@ describe("Faucet", () => {
   });
 
   describe("send", () => {
-    it("can send", async () => {
+    it("can send bank token", async () => {
```
Comment (ethanfrey) [inline packages/faucet/src/faucet.spec.ts]: So nice..

File: scripts/cosm/deploy_erc20.js (hunk 1 above)
Comment (webmaster128) [reply scripts/cosm/deploy_erc20.js]: Yeah I used Uint64 before but then had to change to to arbitrary length BN

File: packages/bcp/src/cosmwasmcodec.spec.ts (hunk 2 above)
Comment (webmaster128) [reply packages/bcp/src/cosmwasmcodec.spec.ts]: For those kind of encoder tests I prefer to be specific to allow seeing what is going on and to reduce dependencies as much as possible. The example in "can send ERC20 tokens" at least removes the fee noise.

File: packages/bcp/src/cosmwasmconnection.ts (hunk 4 above)
Comment (webmaster128) [reply packages/bcp/src/cosmwasmconnection.ts]: The bigger problem I see there is the lack of a proper token identifier. Optimizing working code is relatively easy compared to writing a cross chain spec.

File: packages/faucet/src/faucet.spec.ts (hunk 6 above)
Comment (webmaster128) [reply packages/faucet/src/faucet.spec.ts]: Oh, I mixed up the titles. This is the contract test, not the bank test

File: packages/bcp/src/cosmwasmcodec.spec.ts (hunk 2 above)
Comment (ethanfrey) [reply packages/bcp/src/cosmwasmcodec.spec.ts]: Good call. This is encoding. Totally right, j was thinking usage, and it did seem less noisy

**Pull request merged**